tests/server.py

```diff
@@ -307,16 +308,28 @@ def connectTCP(self, host, port, factory, timeout=30, bindAddress=None):
         return conn
 
     def advance(self, amount):
-        # run any "callFromThread" callbacks before anything registered with
-        # "callLater"
+        # first advance our reactor's time, and run any "callLater" callbacks that
+        # makes ready
+        super().advance(amount)
+
+        # now run any "callFromThread" callbacks
         while True:
             try:
                 callback = self._thread_callbacks.popleft()
             except IndexError:
                 break
             callback()
 
-        super().advance(amount)
+            # check for more "callLater" callbacks added by the thread callback
+            # This isn't required in a regular reactor, but it ends up meaning that
+            # our database queries can complete in a single call to `advance` [1] which
+            # simplifies tests.
+            #
+            # [1]: we replace the threadbool backing the db connection pool with a
+            # mock ThreadPool which doesn't really use threads; but still use
```

Comment:
```suggestion
            # mock ThreadPool which doesn't really use threads; but still uses
```

Reply:
no, it's not the threadpool that does that.

```suggestion
            # mock ThreadPool which doesn't really use threads; but we still use
```